Use post_type parameter to filter email-related templates [MAILPOET-6279]

costasovo opened the pull request:
## Description

This PR improves the fetching of templates for the email editor by moving the filtering of the templates from the client to the REST API.

## Code review notes

Please verify that the template selector offers the correct templates.

## QA notes

We can skip QA.

## Linked PRs

_N/A_

## Linked tickets

[MAILPOET-6279]

## After-merge notes

_N/A_

## Tasks

- [ ] I followed [best practices](https://codex.wordpress.org/I18n_for_WordPress_Developers) for translations
- [ ] I added sufficient test coverage
- [ ] I embraced TypeScript by either creating new files in TypeScript or converting existing JavaScript files when making changes


[MAILPOET-6279]: https://mailpoet.atlassian.net/browse/MAILPOET-6279?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

## Preview

[Preview in WordPress Playground](https://account.mailpoet.com/playground/new/branch:email-editor/templates-fetch-by-post-type)

_The latest successful build from `email-editor/templates-fetch-by-post-type` will be used. If none is available, the link won't work._

costasovo committed:
Use post_type parameter to filter email-related templates

This commit utilizes the post_type parameter that ensures that the REST API returns
only the templates associated with the specified post type.
It improves performance because before the change we were fetching all existing templates and filtering them on client.
[MAILPOET-6279]

lysyjan reviewed (commented):
When I tested it with the WP 6.7 Beta and active theme `Twenty Twenty Five` I could see five templates instead of four. Is it correct?

![Screenshot on 2024-10-16 at 12-46-32](https://github.com/user-attachments/assets/073926d5-487c-4574-993e-1f34acb83b8b)

costasovo commented:
> When I tested it with the WP 6.7 Beta and active theme Twenty Twenty Five I could see five templates instead of four. Is it correct?

@lysyjan It is not expected :( It looks like the template is somehow added to the response, even though the API should filter it. Good catch! I'll dig into it. Maybe we will need to keep the client-side filtering in place to be sure.

costasovo committed:
Reintroduce the client-side filtering of email templates

We still need to filter the templates because, in some cases, the API also returns custom templates
ignoring the post_type filter in the query
[MAILPOET-6279]

costasovo commented:
@lysyjan It looks like there is a bug in the WP's REST API for templates. When a custom template is created in the site editor the API returns it even when the post_type filter is set to `mailpoet_emails`. I returned the client side filtering. I think we should still use the query filter as it still filters a bunch of templates.
